tests/test_photon_array.py

```diff
@@ -243,6 +233,69 @@ def test_photon_array():
     # Error if it doesn't fit.
     assert_raises(ValueError, pa2.assignAt, 90, pa1)
 
+    # Check copyFrom
+    pa2 = galsim.PhotonArray(100)
+    pa2.copyFrom(pa1, slice(0,50))
+    pa2.copyFrom(pa1, target_indices=slice(50,100), source_indices=slice(49,None,-1))
+    np.testing.assert_array_equal(pa2.x[:50], pa1.x)
+    np.testing.assert_array_equal(pa2.y[:50], pa1.y)
+    np.testing.assert_array_equal(pa2.flux[:50], pa1.flux)
+    np.testing.assert_array_equal(pa2.dxdz[:50], pa1.dxdz)
+    np.testing.assert_array_equal(pa2.dydz[:50], pa1.dydz)
+    np.testing.assert_array_equal(pa2.wavelength[:50], pa1.wavelength)
+    np.testing.assert_array_equal(pa2.pupil_u[:50], pa1.pupil_u)
+    np.testing.assert_array_equal(pa2.pupil_v[:50], pa1.pupil_v)
+    np.testing.assert_array_equal(pa2.time[:50], pa1.time)
+    np.testing.assert_array_equal(pa2.x[50:], pa1.x[::-1])
+    np.testing.assert_array_equal(pa2.y[50:], pa1.y[::-1])
+    np.testing.assert_array_equal(pa2.flux[50:], pa1.flux[::-1])
+    np.testing.assert_array_equal(pa2.dxdz[50:], pa1.dxdz[::-1])
+    np.testing.assert_array_equal(pa2.dydz[50:], pa1.dydz[::-1])
+    np.testing.assert_array_equal(pa2.wavelength[50:], pa1.wavelength[::-1])
+    np.testing.assert_array_equal(pa2.pupil_u[50:], pa1.pupil_u[::-1])
+    np.testing.assert_array_equal(pa2.pupil_v[50:], pa1.pupil_v[::-1])
+    np.testing.assert_array_equal(pa2.time[50:], pa1.time[::-1])
+
+    # Can copy a single photon if desired.
+    pa2.copyFrom(pa1, 17, 20)
+    assert pa2.flux[17] == pa1.flux[20]
+    assert pa2.x[17] == pa1.x[20]
+    assert pa2.time[17] == pa1.time[20]
+
+    # Can choose not to copy flux
+    pa2.flux[27] = -1
+    pa2.copyFrom(pa1, 27, 10, do_flux=False)
+    assert pa2.flux[27] != pa1.flux[10]
+    assert pa2.x[27] == pa1.x[10]
+    assert pa2.time[27] == pa1.time[10]
+
+    # ... or positions
+    pa2.copyFrom(pa1, 37, 8, do_xy=False)
+    assert pa2.flux[37] == pa1.flux[8]
+    assert pa2.x[37] != pa1.x[8]
+    assert pa2.y[37] != pa1.y[8]
+    assert pa2.time[37] == pa1.time[8]
+
+    # ... or the other arrays
+    pa2.dxdz[47] = pa2.dydz[47] = pa2.wavelength[47] = -1
+    pa2.pupil_u[47] = pa2.pupil_v[47] = pa2.time[47] = -1
+    pa2.copyFrom(pa1, 47, 18, do_other=False)
+    assert pa2.flux[47] == pa1.flux[18]
+    assert pa2.x[47] == pa1.x[18]
+    assert pa2.y[47] == pa1.y[18]
+    assert pa2.dxdz[47] != pa1.dxdz[18]
+    assert pa2.dydz[47] != pa1.dydz[18]
+    assert pa2.wavelength[47] != pa1.wavelength[18]
+    assert pa2.pupil_u[47] != pa1.pupil_u[18]
+    assert pa2.pupil_v[47] != pa1.pupil_v[18]
+    assert pa2.time[47] != pa1.time[18]
+
+    # Error if indices are invalid
+    assert_raises(ValueError, pa2.copyFrom, pa1, slice(50,None), slice(50,None))
+    assert_raises(ValueError, pa2.copyFrom, pa1, 100, 0)
+    assert_raises(ValueError, pa2.copyFrom, pa1, 0, slice(None))
+    assert_raises(ValueError, pa2.copyFrom, pa1)
+
```

Comment:
Should probably add some tests for numpy fancy indexing too.
I.e., `pa1.copyFrom(pa2, pa2.flux>10)` and `pa1.copyFrom(pa2, np.where(pa2.flux>10))`

Reply:
And maybe even discontiguous slices.  `pa1.copyFrom(pa2, slice(0, 100, 2))`

Reply:
Yep.  Good ideas.  Will do.